File: mgyoutube-multitech/api-node/src/modules/defaultvideomoduleimpl.ts
import { Video } from '../apimodel/video';
import { VideoModule } from './videomodule';
import { google, youtube_v3 } from 'googleapis'


export class DefaultVideoModuleImpl implements VideoModule {
    private apiKey: string;
    private applicationName: string

    constructor(apiKey: string, applicationName: string) {
        this.apiKey = apiKey;
        this.applicationName = applicationName;
    }

    // setUserDataRepo(udr: UserDataRepo): void {
    //     this.userDataRepo = udr;
    // }

    private getYouTube(): youtube_v3.Youtube {
        var youtube = google.youtube({
            version: 'v3',
            auth: this.apiKey
        });
        return youtube;
    }

    public async search(searchTerms: string): Promise<Array<Video>> {

        var youtube = this.getYouTube();

        var searchResults = await youtube.search.list({
            part: 'id,snippet',
            q: searchTerms,
            safeSearch: 'strict',
            type: 'video',
            maxResults: 50
        });

        var videos: Array<Video> = [];

        if (searchResults.status == 200) {
            if (searchResults.data.items) {
                // the "for (var searchResult in searchResults.data.items)" loop isn't typing correctly so doing it the old fashion way
                for (var i = 0; i < searchResults.data.items.length; i++) {
                    var searchResult = searchResults.data.items[0];
                    var video = new Video();
                    if (searchResult.snippet) {
                        if (searchResult.id) {
                            video.videoId = searchResult.id.videoId
                        }
                        video.title = searchResult.snippet.title;
                        video.description = searchResult.snippet.description;
                        video.channelTitle = searchResult.snippet.channelTitle
                        video.channelId = searchResult.snippet.channelId
                        video.publishedAt = searchResult.snippet.publishedAt
                        if (searchResult.snippet.thumbnails && searchResult.snippet.thumbnails.default) {
                            video.thumbnailUrl = searchResult.snippet.thumbnails.default.url
                            video.thumbnailWidth = searchResult.snippet.thumbnails.default.width
                            video.thumbnailHeight = searchResult.snippet.thumbnails.default.height
                        }
                    }
                    videos.push(video);
                }
            }
            else {
                console.warn("searchResults status indicates failure: ", searchResults.status)
            }
        }

        return videos;
    }
}

File: mgyoutube-multitech/api-python/modules/DefaultVideoModuleImpl.py
from apimodel import Video
from modules.VideoModule import VideoModule
from googleapiclient.discovery import build, Resource


class DefaultVideoModuleImpl(VideoModule):
    YOUTUBE_API_SERVICE_NAME = "youtube"
    YOUTUBE_API_VERSION = "v3"
    YOUTUBE_API_MAX_VIDEOS_RETURNED = 50

    def __init__(self, apiKey: str, applicationName: str):
        self.apiKey = apiKey
        self.applicationName = applicationName

    def search(self, searchTerms: str) -> list:
        youtube = build(DefaultVideoModuleImpl.YOUTUBE_API_SERVICE_NAME,
                        DefaultVideoModuleImpl.YOUTUBE_API_VERSION, developerKey=self.apiKey)

        #
        # false positve pylink(no-member) "Instance of 'Resource' has no 'search' member"
        # dont want to disable it in general
        #
        response = youtube.search().list(
            q=searchTerms,
            part='id,snippet',
            maxResults=DefaultVideoModuleImpl.YOUTUBE_API_MAX_VIDEOS_RETURNED,
            type="video",
            safeSearch="strict"
        ).execute()

        # print("search '", searchTerms, "' returned:", response.items)

        videos = list()

        # type(response.items) <class 'builtin_function_or_method'>
        # print("type(response.items)", type(response.items))
        # response.items <built-in method items of dict object at 0x000001E821D51A98>
        # print("response.items", response.items)
        # response.items() dict_items([('kind', 'youtube#searchListResponse'),...
        # print("response.items()", response.items())

        #
        # WHY?!?! if response.items is a dict() why can we not subscript it or call get()
        #
        # TypeError: argument of type 'builtin_function_or_method' is not iterable
        # print("items in response.items", 'items' in response.items)
        # TypeError: 'builtin_function_or_method' object is not subscriptable
        # print("response.items['items']", response.items['items'])
        # AttributeError: 'builtin_function_or_method' object has no attribute 'get'
        # print("response.items.get('items')", response.items.get('items'))

        for responseItemKey, responseItem in response.items():
            # print("responseItemKey", responseItemKey)
            if (responseItemKey == "items"):
                # print("type(responseItem)", type(responseItem))
                # print("len(responseItem)", len(responseItem))
                videoResult = responseItem[0]
                #print("type(videoResult)", type(videoResult))
                #print("videoResult", videoResult)
                for videoResult in responseItem:
                    id = videoResult['id']
                    #print("type(id)", type(id))
                    #print("id", id)
                    snippet = videoResult['snippet']
                    #print("type(snippet)", type(snippet))
                    thumbnails = snippet['thumbnails']
                    #print("type(thumbnails)", type(thumbnails))
                    #print("thumbnails", thumbnails)
                    video = Video.Video()
                    video.videoId = id['videoId']
                    video.title = snippet['title']
                    video.description = snippet['description']
                    video.channelId = snippet['channelId']
                    video.channelTitle = snippet['channelTitle']
                    video.thumbnailUrl = thumbnails['default']['url']
                    video.thumbnailHeight = thumbnails['default']['height']
                    video.thumbnailWidth = thumbnails['default']['width']
                    video.publishedAt = snippet['publishedAt']

                    videos.append(video)

        return videos
